### PR TITLE
Use int64_t for lengths

### PR DESCRIPTION
**Pull request by DemiMarie**
This prevents an integer overflow in dataLength().

### CODE REVIEW

**Comment by pmatilai** [comment]
The added check may well be right, but it hardly prevents "any possible overflow" there could ever be.

If you look at past commits, we go to great lengths to explain in which situation some problem can happen and how the existing checks are not suffient, please help keep the tradition. It helps the reviewer to check the code does what you think it does (this is subtle and treacherous code, been there) and seeing whether there's a better way to do that (see the string length check PR for example)

Last but certainly not least, *when* (not if) somebody later on stumbles on another bug in the surrounding code, a detailed explanation as to why a piece of code is the way it is helps avoid breaking that subtle case again. Etc. This goes to most of these hardening commits - some things are plain obvious but others are not.

**Comment by pmatilai** [comment]
"int" is (at least) 32 bits on all platforms rpm supports, that's no. The rough idea is that explicitly sized types are used for "physical" data of that size, and abstract types for abstractions, such as results of calculation.

In fact this (and various other similar places) would probably be better off using a 64bit type so it's not possible to overflow it with the data we're dealing with. Overflow checks all over the place hurt readability, it's better to eliminate the problem entirely if at all possible.

**Comment by DemiMarie** [comment]
@pmatilai does this revised version look good?

File: lib/header.c [hunk 1]
```diff
@@ -1863,13 +1862,7 @@ static rpmRC hdrblobVerifyRegion(rpmTagVal regionTag, int exact_size,
...
     }
 
-    /*
-     * Trailer offset is negative and has a special meaning.  Be sure to negate
-     * *after* the division, so the negation cannot overflow.  The parentheses
-     * around the division are required!
-     *
-     * Thankfully, the modulus operator works fine on negative numbers.
-     */
+    /* Negate after division to avoid overflow.  Watch out for precedence! */
```
Comment (pmatilai) [inline lib/header.c:1865]: If you want to change this comment, put it into a separate commit (no need for separate PR though). Having it here makes it look like it's somehow related to this code change, when it isn't.

**Comment by pmatilai** [comment]
All these casts are making my skin crawl. My golden rule is that needing a cast is a warning sign that says to look for another way. I'm in a bit of denial over the need to cast to int64 to ensure the calculation occurs at the target variable type - are our "smart" optimizing compilers *really* that dumb? I get the need for such a cast in eg anonymous conditional context, but not when the target variable type is well known such as here.

Also the cast on the return is unnecessary, the compiler will do the right thing without our help so let it. The INT_MAX check makes the point clear enough.

**Comment by DemiMarie** [comment]
> All these casts are making my skin crawl. My golden rule is that needing a cast is a warning sign that says to look for another way. I'm in a bit of denial over the need to cast to int64 to ensure the calculation occurs at the target variable type - are todays "smart" optimizing compilers _really_ that dumb? I get the need for such a cast in eg anonymous conditional context, but not when the target variable type is well known such as here.

Sadly it is necessary :disappointed:.  That said, there is another option: use GCC and Clang’s overflow-checking builtins, such as `__builtin_mul_overflow`.  Those always have well-defined behavior and report overflow, which is what we really want IMO.

Admittedly, `__builtin_mul_overflow` isn’t part of any C standard (though it should be, IMO).  However, we use GCC/Clang extensions elsewhere in the code, so I would be fine with this.

> Also the cast on the return is unnecessary, the compiler will do the right thing without our help so let it. The INT_MAX check makes the point clear enough.

That’s fine.

**Comment by DemiMarie** [comment]
I found a cleaner solution: assign `count` to an `int64_t` variable.

File: lib/header.c [hunk 2]
```diff
@@ -467,15 +468,14 @@ static int dataLength(rpm_tagtype_t type, rpm_constdata_t p, rpm_count_t count,
 	break;
 
     default:
-	if (typeSizes[type] == -1)
-	    return -1;
-	length = typeSizes[(type & 0xf)] * count;
-	if (length < 0 || (se && (s + length) > se))
+	/* this won’t overflow, since ‘long_count’ is 64-bit */
+	length = typeSizes[type] * long_count;
+	if (length <= 0 || (se && length > se - s))
```
Comment (pmatilai) [inline lib/header.c:473]: I don't see a need for an extra variable. You could just assign to length and multiply, or make eg typeSizes int64 type (and/or count itself for that matter)

Note that this ends up removing an old safety feature in the code, ie that `(type & 0xf)` ensures that we'll never access typeSizes out of bounds. Only there was that redundant `== -1` check before it, making it moot :facepalm:

Comment (DemiMarie) [reply lib/header.c:473]: I went with making `typeSizes` 64-bit, which will help prevent similar problems elsewhere.

**Comment by pmatilai** [comment]
Not sure if the earlier versions had those comments in dataLength(), if they did then apologies for missing. Scratch them, otherwise we'll end up having a comment on every other row of the code. A simple remark like "note use of 64bit types to avoid overflow" before the function body is more than enough.

However I have a growing uneasy feeling that we've long since gotten lost in the weeds.

We're talking about data that is capped to 256MB total. We're *nowhere near* the overflow points of 32bit ints, whether signed or not. So actually the direction of #1493 is probably a far saner approach, only the rationale is backwards and misses the actual point (and so I missed it too): the check is in the wrong place entirely. regionSwab() is "inner sanctum" already and shouldn't have to check the data type basics, that is already done earlier, except that it's not for the count at the moment.

**Comment by pmatilai** [comment]
So yeah, I'll note this here now for lack of better place.

*I* am overflowing with all these PR's to add a little check here and another on there. Time to take a step back and cool off a little. There clearly are various missing checks in the code but seems to me we're running in circles putting out small fires here and there rather than addressing the fundamental issues (already hinted at this in #1494)

**Comment by DemiMarie** [comment]
> So yeah, I'll note this here now for lack of better place.
> 
> _I_ am overflowing with all these PR's to add a little check here and another on there. Time to take a step back and cool off a little. There clearly are various missing checks in the code but seems to me we're running in circles putting out small fires here and there rather than addressing the fundamental issues (already hinted at this in #1494)

The biggest question is _what the fundamental issues are_.  Using signed types where unsigned types should be used is one.  Not checking for overflow before arithmetic is another.  That said, I believe the biggest improvements are:

- Continuous fuzzing.  libfuzzer is incredibly good at finding bugs, especially when combined with ASAN and UBSAN.  And once we get it set up, we can enroll rpm in Google’s OSS-Fuzz, which will perform automatic fuzzing on every commit at massive scale.
- Use a bounded buffer abstraction for parsing and serialization.  That lets us write the security-critical bounds (and other) checks _once_, and centralizes them where they can be easily audited and tested.  BoringSSL has such an abstraction under an ISC license already.
- Clearly separate untrusted and trusted data.  QubesOS uses variables with names starting with `untrusted_` for the former.
- More tests.  We probably won’t be able to go as far as SQLite, but we can go farther than we do now.
- Be as strict as possible when parsing.  While we are limited by backwards compatibility, we should be as strict as possible within that constraint.  RPMv6 is an opportunity to tighten the rules on what will be accepted, and I suggest using it.

Some longer-term ideas:

- Consider alternatives to C.  I understand that Rust is not currently portable enough, as RPM supports architectures that LLVM does not.  That said, a GCC backend for Rust is in development, which will make Rust able to target any platform GCC supports.  More immediately, we might be able to move some logic to Lua.  If we required SQLite, we might even be able to implement some of the logic as SQL queries against the rpmdb.
- (Someday) Formal methods.  Just being able to prove that certain functions (such as `headerImport`) don’t have undefined behavior would be awesome.

**Comment by pmatilai** [comment]
> The biggest question is what the fundamental issues are. 

Yes, identifying those is where it needs to begin.

Much like, if your house is swarmed with mosquitos, first you identify where they're getting in, close the window and only then commence at swatting the remaining ones :sweat_smile: 

The existing header/package code is a chainsaw job from consolidating something like four different codepaths that did nearly the same thing, but each with different faults, and then bolt on the enforcing policy stuff in a mad sprint to meet deadlines. While it does *much* more checking than it previously did, there are necessarily many rough cuts and missed opportunities to clean up and improve in there.
